File: packages/common/erp-common-expressjs/src/middlewares/authorization.middleware.ts
import { Request, Response, NextFunction } from 'express';
import { Context, sendError, TypeErrorMessage, Winston } from '@gmahechas/erp-common-ms-utils-js';

export const authorizationMiddleware = (scopes: string[]) => async (request: Request, response: Response, next: NextFunction) => {
	const auth = Context.get('auth');
	const { body, originalUrl, method, params, query } = request;
	const scope = JSON.parse(auth.scope);
	for (const capability of scopes) {
		const [service, actions] = capability.split(':');
		if (!scope[service]) {
			Winston.logger.error(TypeErrorMessage.AUTHORIZATION, { requestId: Context.get('requestId'), auth: JSON.stringify(auth), action: originalUrl, method, payload: JSON.stringify({ body, params, query }) });
			sendError(TypeErrorMessage.AUTHORIZATION);
		}
		for (const action of actions.split(',')) {
			if (!scope[service].includes(action)) {
				Winston.logger.error(TypeErrorMessage.AUTHORIZATION, { requestId: Context.get('requestId'), auth: JSON.stringify(auth), action: originalUrl, method, payload: JSON.stringify({ body, params, query }) });
				sendError(TypeErrorMessage.AUTHORIZATION);
			}
		}
	}
	next();
};

File: packages/common/erp-common-flaskpy/gmahechas_erp_common_flaskpy/middlewares/authorization_middleware.py
from functools import wraps
from flask import Response
import json
from gmahechas_erp_common_py import response
from gmahechas_erp_common_ms_utils_py import Context


def authorization_middleware(scopes):
    @wraps(scopes)
    def authorization_middleware_wrap(function):
        @wraps(function)
        def wrapper(*args, **kwargs):
            auth = Context.get("auth")
            scope = json.loads(auth["scope"])
            for capability in scopes:
                [service, actions] = capability.split(":")
                if not scope[service]:
                    data = response(400, error={"type": "authorization"})
                    status_code, body = data['statusCode'], data['body']
                    return Response(body, mimetype='application/json', status=status_code)

                for action in actions.split(":"):
                    if action not in scope[service]:
                        data = response(400, error={"type": "authorization"})
                        status_code, body = data['statusCode'], data['body']
                        return Response(body, mimetype='application/json', status=status_code)

            return function(*args, **kwargs)

        return wrapper

    return authorization_middleware_wrap
